etc/convert-test-results.swift

```diff
@@ -0,0 +1,332 @@
+            throw ParsingError("unable to parse TimeInterval from \(input)")
+        }
+        self = time
+    }
+}
+
+func ensureSuiteMatches(old: String, new: String) throws {
+    guard old == new else {
+        throw ParsingError("test suite name \(new) does not match previously found name for current suite \(old)")
+    }
+}
+
+/// State machine which processes `swift test` output and updates itself accordingly.
+enum ParsingState {
+    /// In the following cases:
+    /// - `completeTests` is stored whenever we are in the middle of a suite, and contains any test cases we have found
+    ///   for the suite so far.
+    /// - `completeSuites` is always stored and contains all of the suites we have fully parsed so far.
+
+    /// We have read a line indicating that a test suite with the given name started, but we are not in a particular
+    /// test case.
+    case inSuite(name: String, completeTests: [TestCase], completeSuites: [TestSuite])
+    /// We've read a line indicating that a suite with the given name ended, and are expecting to read a next line
+    /// containing test pass/fail counts for the suite.
+    case awaitingSuiteDetails(name: String, completeTests: [TestCase], completeSuites: [TestSuite])
+    /// We are in the middle of a test with the given name in the given suite. `output` contains any output the test
+    /// case has produced.
+    case inTest(suite: String, name: String, output: [String], completeTests: [TestCase], completeSuites: [TestSuite])
+    /// None of the above. This means we are between test suites.
+    case none(completeSuites: [TestSuite])
+
+    // account for variation in formatting of test output on platforms. this assumes you are running this script on the
+    // same platform where you ran the tests.
+#if os(macOS)
+    static let testCaseStartedRegex = try! NSRegularExpression(pattern: #"Test Case '-\[.+\.(.+) (.+)\]' started"#)
+    static let testCaseStatusRegex = try! NSRegularExpression(pattern:#"Test Case '-\[.+\.(.+) (.+)\]' (passed|failed) \((.+) seconds\)"#)
+#else
+    static let testCaseStartedRegex = try! NSRegularExpression(pattern:#"Test Case '(.+)\.(.+)' started"#)
+    static let testCaseStatusRegex =  try! NSRegularExpression(pattern:#"Test Case '(.+)\.(.+)' (passed|failed) \((.+) seconds\)"#)
+#endif
+    static let testSuiteStartedRegex = try! NSRegularExpression(pattern:#"Test Suite '(.+)' started"#)
+    static let testSuiteStatusRegex = try! NSRegularExpression(pattern:#"Test Suite '(.+)' (passed|failed)"#)
+    static let testSuiteDetailsRegex = try! NSRegularExpression(pattern:#"Executed (\d+) tests?, with (\d+) failures? \((\d+) unexpected\) in (.+) \("#)
+
+    /// Processes a new line of test output and updates self accordingly.
+    mutating func processLine(_ line: String) throws {
+        let fullRange = NSRange(line.startIndex..<line.endIndex, in: line)
+
+        if let match = Self.testCaseStartedRegex.firstMatch(in: line, range: fullRange) {
```

Comment:
I tried to store a map of regex: handler function here so this could be done with a for loop rather than separate statements, but the compiler did not like my attempting to store mutating instance methods in a dictionary. I was running into the error described [here](https://stackoverflow.com/questions/37488316/partial-application-of-mutating-method-is-not-allowed), which I had not seen before but is a rather interesting case.